File: backend/code/sdk/javascript/ai-trading-sdk.ts
/**
 * AI Trading Evolution - JavaScript/TypeScript SDK
 * ================================================
 * Official JavaScript client library for AI Trading Evolution API
 * 
 * @author MiniMax Agent
 * @version 1.0.0
 * @date 2025-11-04
 * 
 * @example
 * ```javascript
 * import { TradingClient } from 'ai-trading-sdk';
 * 
 * const client = new TradingClient({
 *   baseUrl: 'http://localhost:8000',
 *   apiKey: 'your-api-key'
 * });
 * 
 * // Get market data
 * const data = await client.market.getData('BTC/USDT');
 * 
 * // Execute strategy
 * const signal = await client.strategy.execute('grid', 'BTC/USDT');
 * 
 * // Run analytics
 * const sentiment = await client.analytics.sentiment('BTC/USDT');
 * ```
 */

// =============================================================================
// Types
// =============================================================================

export interface ClientConfig {
  baseUrl?: string;
  apiKey?: string;
  timeout?: number;
}

export interface MarketDataRequest {
  symbol: string;
  market_type?: 'crypto' | 'forex' | 'stocks' | 'commodities';
  timeframe?: string;
  limit?: number;
}

export interface StrategyRequest {
  strategy_name: string;
  symbol: string;
  timeframe?: string;
  parameters?: Record<string, any>;
}

export interface AnalyticsRequest {
  analysis_type: string;
  symbol?: string;
  parameters?: Record<string, any>;
}

export interface HealthResponse {
  status: string;
  timestamp: string;
  version: string;
  uptime: number;
  modules: Record<string, any>;
}

export interface StrategyResponse {
  strategy_name: string;
  symbol: string;
  signal: 'BUY' | 'SELL' | 'HOLD';
  confidence: number;
  price: number;
  entry_price?: number;
  stop_loss?: number;
  take_profit?: number;
  timestamp: string;
  metadata: Record<string, any>;
}

export interface WebSocketMessage {
  action: 'subscribe' | 'unsubscribe' | 'ping' | 'get_stats';
  channel?: string;
}

// =============================================================================
// API Error
// =============================================================================

export class APIError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: any
  ) {
    super(message);
    this.name = 'APIError';
  }
}

// =============================================================================
// Base Client
// =============================================================================

class BaseClient {
  private baseUrl: string;
  private apiKey?: string;
  private timeout: number;

  constructor(config: ClientConfig = {}) {
    this.baseUrl = (config.baseUrl || 'http://localhost:8000').replace(/\/$/, '');
    this.apiKey = config.apiKey;
    this.timeout = config.timeout || 30000;
  }

  private async request<T>(
    method: string,
    endpoint: string,
    data?: any,
    params?: Record<string, string>
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${endpoint}`);
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        url.searchParams.append(key, value);
      });
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url.toString(), {
        method,
        headers,
        body: data ? JSON.stringify(data) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const error = await response.json().catch(() => ({ error: 'Unknown error' }));
        throw new APIError(
          `API Error ${response.status}: ${error.error || 'Unknown error'}`,
          response.status,
          error
        );
      }

      return await response.json();
    } catch (error) {
      if (error instanceof APIError) {
        throw error;
      }
      if (error.name === 'AbortError') {
        throw new APIError('Request timeout');
      }
      throw new APIError(`Connection error: ${error.message}`);
    }
  }

  async get<T>(endpoint: string, params?: Record<string, string>): Promise<T> {
    return this.request<T>('GET', endpoint, undefined, params);
  }

  async post<T>(endpoint: string, data?: any): Promise<T> {
    return this.request<T>('POST', endpoint, data);
  }

  getBaseUrl(): string {
    return this.baseUrl;
  }
}

// =============================================================================
// Market API
// =============================================================================

export class MarketAPI {
  constructor(private client: BaseClient) {}

  /**
   * Get market data
   */
  async getData(
    symbol: string,
    options: Partial<MarketDataRequest> = {}
  ): Promise<any> {
    return this.client.post('/api/v1/market/data', {
      symbol,
      market_type: options.market_type || 'crypto',
      timeframe: options.timeframe || '1h',
      limit: options.limit || 100,
    });
  }

  /**
   * Get list of available symbols
   */
  async listSymbols(marketType: string = 'crypto'): Promise<any> {
    return this.client.get('/api/v1/market/symbols', { market_type: marketType });
  }
}

// =============================================================================
// Strategy API
// =============================================================================

export class StrategyAPI {
  constructor(private client: BaseClient) {}

  /**
   * Execute trading strategy
   */
  async execute(
    strategyName: string,
    symbol: string,
    options: Partial<StrategyRequest> = {}
  ): Promise<StrategyResponse> {
    return this.client.post<StrategyResponse>('/api/v1/strategy/execute', {
      strategy_name: strategyName,
      symbol,
      timeframe: options.timeframe || '1h',
      parameters: options.parameters || {},
    });
  }

  /**
   * Get list of available strategies
   */
  async list(): Promise<any> {
    return this.client.get('/api/v1/strategy/list');
  }
}

// =============================================================================
// Analytics API
// =============================================================================

export class AnalyticsAPI {
  constructor(private client: BaseClient) {}

  /**
   * Run analytics
   */
  async analyze(
    analysisType: string,
    symbol?: string,
    parameters?: Record<string, any>
  ): Promise<any> {
    return this.client.post('/api/v1/analytics/analyze', {
      analysis_type: analysisType,
      symbol,
      parameters: parameters || {},
    });
  }

  /**
   * Get sentiment analysis
   */
  async sentiment(symbol: string): Promise<any> {
    return this.analyze('sentiment', symbol);
  }

  /**
   * Get risk scoring
   */
  async riskScoring(symbol: string): Promise<any> {
    return this.analyze('risk_scoring', symbol);
  }

  /**
   * Get list of analytics types
   */
  async listTypes(): Promise<any> {
    return this.client.get('/api/v1/analytics/types');
  }
}

// =============================================================================
// WebSocket Client
// =============================================================================

export class WebSocketClient {
  private ws?: WebSocket;
  private clientId?: string;
  private subscriptions: Set<string> = new Set();
  private messageHandlers: Map<string, Function[]> = new Map();

  constructor(private baseUrl: string) {
    this.baseUrl = baseUrl.replace('http://', 'ws://').replace('https://', 'wss://');
  }

  /**
   * Connect to WebSocket
   */
  async connect(clientId?: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.clientId = clientId || this.generateClientId();
      const url = `${this.baseUrl}/ws/${this.clientId}`;

      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
      };

      this.ws.onmessage = (event) => {
        const message = JSON.parse(event.data);
        
        if (message.type === 'welcome') {
          resolve();
        }
        
        // Trigger handlers
        const handlers = this.messageHandlers.get(message.type) || [];
        handlers.forEach(handler => handler(message));
        
        // Trigger channel handlers
        if (message.channel) {
          const channelHandlers = this.messageHandlers.get(message.channel) || [];
          channelHandlers.forEach(handler => handler(message));
        }
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
        reject(error);
      };

      this.ws.onclose = () => {
        console.log('WebSocket disconnected');
      };
    });
  }

  /**
   * Subscribe to channel
   */
  async subscribe(channel: string, handler?: Function): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    this.send({
      action: 'subscribe',
      channel,
    });

    this.subscriptions.add(channel);

    if (handler) {
      this.on(channel, handler);
    }
  }

  /**
   * Unsubscribe from channel
   */
  async unsubscribe(channel: string): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    this.send({
      action: 'unsubscribe',
      channel,
    });

    this.subscriptions.delete(channel);
    this.messageHandlers.delete(channel);
  }

  /**
   * Send message
   */
  send(message: WebSocketMessage): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    this.ws.send(JSON.stringify(message));
  }

  /**
   * Register message handler
   */
  on(event: string, handler: Function): void {
    if (!this.messageHandlers.has(event)) {
      this.messageHandlers.set(event, []);
    }
    this.messageHandlers.get(event)!.push(handler);
  }

  /**
   * Send ping
   */
  async ping(): Promise<void> {
    this.send({ action: 'ping' });
  }

  /**
   * Close connection
   */
  close(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = undefined;
    }
  }

  private generateClientId(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }
}

// =============================================================================
// Main Trading Client
// =============================================================================

/**
 * Main AI Trading Evolution client
 * 
 * @example
 * ```typescript
 * const client = new TradingClient({
 *   baseUrl: 'http://localhost:8000',
 *   apiKey: 'your-api-key'
 * });
 * 
 * // Get market data
 * const data = await client.market.getData('BTC/USDT');
 * 
 * // Execute strategy
 * const signal = await client.strategy.execute('grid', 'BTC/USDT');
 * 
 * // Run analytics
 * const sentiment = await client.analytics.sentiment('BTC/USDT');
 * 
 * // WebSocket
 * const ws = client.websocket();
 * await ws.connect();
 * await ws.subscribe('market:BTC/USDT', (message) => {
 *   console.log('Market update:', message);
 * });
 * ```
 */
export class TradingClient {
  private _baseClient: BaseClient;

  public readonly market: MarketAPI;
  public readonly strategy: StrategyAPI;
  public readonly analytics: AnalyticsAPI;

  constructor(config: ClientConfig = {}) {
    this._baseClient = new BaseClient(config);

    // Initialize API endpoints
    this.market = new MarketAPI(this._baseClient);
    this.strategy = new StrategyAPI(this._baseClient);
    this.analytics = new AnalyticsAPI(this._baseClient);
  }

  /**
   * Get API health status
   */
  async health(): Promise<HealthResponse> {
    return this._baseClient.get<HealthResponse>('/health');
  }

  /**
   * Get performance metrics
   */
  async metrics(): Promise<any> {
    return this._baseClient.get('/metrics');
  }

  /**
   * Get WebSocket client
   */
  websocket(): WebSocketClient {
    return new WebSocketClient(this._baseClient.getBaseUrl());
  }
}

// =============================================================================
// Export all
// =============================================================================

export default TradingClient;

File: backend/sdk/python/ai_trading_sdk.py
"""
AI Trading Evolution - Python SDK
==================================
Official Python client library for AI Trading Evolution API

Author: MiniMax Agent
Version: 1.0.0
Date: 2025-11-04

Installation:
    pip install ai-trading-sdk

Usage:
    from ai_trading_sdk import TradingClient
    
    client = TradingClient(api_key="your-api-key")
    
    # Get market data
    data = await client.market.get_data("BTC/USDT")
    
    # Execute strategy
    signal = await client.strategy.execute("grid", "BTC/USDT")
    
    # Run analytics
    sentiment = await client.analytics.sentiment("BTC/USDT")
"""

import asyncio
import aiohttp
import json
from typing import Dict, List, Optional, Any, Union
from datetime import datetime
import websockets


class APIError(Exception):
    """API error exception"""
    pass


class BaseClient:
    """Base API client"""
    
    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_key: Optional[str] = None,
        timeout: int = 30
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = aiohttp.ClientTimeout(total=timeout)
        self.session: Optional[aiohttp.ClientSession] = None
    
    async def __aenter__(self):
        await self.connect()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
    
    async def connect(self):
        """Initialize HTTP session"""
        if not self.session:
            headers = {}
            if self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
            
            self.session = aiohttp.ClientSession(
                headers=headers,
                timeout=self.timeout
            )
    
    async def close(self):
        """Close HTTP session"""
        if self.session:
            await self.session.close()
            self.session = None
    
    async def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """Make HTTP request"""
        if not self.session:
            await self.connect()
        
        url = f"{self.base_url}{endpoint}"
        
        try:
            async with self.session.request(
                method,
                url,
                json=data,
                params=params
            ) as response:
                if response.status >= 400:
                    error_data = await response.json()
                    raise APIError(f"API Error {response.status}: {error_data.get('error', 'Unknown error')}")
                
                return await response.json()
        
        except aiohttp.ClientError as e:
            raise APIError(f"Connection error: {e}")
    
    async def get(self, endpoint: str, params: Optional[Dict] = None) -> Dict[str, Any]:
        """GET request"""
        return await self._request("GET", endpoint, params=params)
    
    async def post(self, endpoint: str, data: Optional[Dict] = None) -> Dict[str, Any]:
        """POST request"""
        return await self._request("POST", endpoint, data=data)


class MarketAPI:
    """Market data API"""
    
    def __init__(self, client: BaseClient):
        self.client = client
    
    async def get_data(
        self,
        symbol: str,
        market_type: str = "crypto",
        timeframe: str = "1h",
        limit: int = 100
    ) -> Dict[str, Any]:
        """
        Get market data
        
        Args:
            symbol: Trading pair (e.g., "BTC/USDT")
            market_type: Market type (crypto, forex, stocks, commodities)
            timeframe: Timeframe (1m, 5m, 15m, 1h, 4h, 1d)
            limit: Number of candles (1-1000)
        
        Returns:
            Market data dict
        """
        return await self.client.post("/api/v1/market/data", {
            "symbol": symbol,
            "market_type": market_type,
            "timeframe": timeframe,
            "limit": limit
        })
    
    async def list_symbols(self, market_type: str = "crypto") -> Dict[str, Any]:
        """
        Get list of available symbols
        
        Args:
            market_type: Market type (crypto, forex, stocks, commodities)
        
        Returns:
            Symbols dict
        """
        return await self.client.get(f"/api/v1/market/symbols?market_type={market_type}")


class StrategyAPI:
    """Trading strategy API"""
    
    def __init__(self, client: BaseClient):
        self.client = client
    
    async def execute(
        self,
        strategy_name: str,
        symbol: str,
        timeframe: str = "1h",
        parameters: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Execute trading strategy
        
        Args:
            strategy_name: Strategy name (arbitrage, grid, dca, futures, mean_reversion, momentum)
            symbol: Trading pair
            timeframe: Timeframe
            parameters: Strategy parameters
        
        Returns:
            Strategy response with signal
        """
        return await self.client.post("/api/v1/strategy/execute", {
            "strategy_name": strategy_name,
            "symbol": symbol,
            "timeframe": timeframe,
            "parameters": parameters or {}
        })
    
    async def list(self) -> Dict[str, Any]:
        """
        Get list of available strategies
        
        Returns:
            Strategies dict
        """
        return await self.client.get("/api/v1/strategy/list")


class AnalyticsAPI:
    """Analytics API"""
    
    def __init__(self, client: BaseClient):
        self.client = client
    
    async def analyze(
        self,
        analysis_type: str,
        symbol: Optional[str] = None,
        parameters: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Run analytics
        
        Args:
            analysis_type: Analysis type (sentiment, whale_tracking, risk_scoring, portfolio)
            symbol: Trading pair (optional)
            parameters: Analysis parameters
        
        Returns:
            Analysis results
        """
        return await self.client.post("/api/v1/analytics/analyze", {
            "analysis_type": analysis_type,
            "symbol": symbol,
            "parameters": parameters or {}
        })
    
    async def sentiment(self, symbol: str) -> Dict[str, Any]:
        """
        Get sentiment analysis
        
        Args:
            symbol: Trading pair
        
        Returns:
            Sentiment data
        """
        return await self.analyze("sentiment", symbol)
    
    async def risk_scoring(self, symbol: str) -> Dict[str, Any]:
        """
        Get risk scoring
        
        Args:
            symbol: Trading pair
        
        Returns:
            Risk metrics
        """
        return await self.analyze("risk_scoring", symbol)
    
    async def list_types(self) -> Dict[str, Any]:
        """
        Get list of analytics types
        
        Returns:
            Analytics types dict
        """
        return await self.client.get("/api/v1/analytics/types")


class WebSocketClient:
    """WebSocket client for real-time data"""
    
    def __init__(self, base_url: str = "ws://localhost:8000"):
        self.base_url = base_url.replace("http://", "ws://").replace("https://", "wss://")
        self.client_id = None
        self.websocket = None
        self.subscriptions: List[str] = []
    
    async def connect(self, client_id: Optional[str] = None):
        """Connect to WebSocket"""
        if client_id:
            self.client_id = client_id
        else:
            import uuid
            self.client_id = str(uuid.uuid4())
        
        url = f"{self.base_url}/ws/{self.client_id}"
        self.websocket = await websockets.connect(url)
        
        # Receive welcome message
        welcome = await self.websocket.recv()
        print(f"Connected: {welcome}")
    
    async def subscribe(self, channel: str):
        """Subscribe to channel"""
        if not self.websocket:
            raise RuntimeError("Not connected")
        
        await self.websocket.send(json.dumps({
            "action": "subscribe",
            "channel": channel
        }))
        
        self.subscriptions.append(channel)
        
        # Wait for confirmation
        response = await self.websocket.recv()
        print(f"Subscribed: {response}")
    
    async def unsubscribe(self, channel: str):
        """Unsubscribe from channel"""
        if not self.websocket:
            raise RuntimeError("Not connected")
        
        await self.websocket.send(json.dumps({
            "action": "unsubscribe",
            "channel": channel
        }))
        
        if channel in self.subscriptions:
            self.subscriptions.remove(channel)
        
        # Wait for confirmation
        response = await self.websocket.recv()
        print(f"Unsubscribed: {response}")
    
    async def receive(self) -> Dict[str, Any]:
        """Receive message"""
        if not self.websocket:
            raise RuntimeError("Not connected")
        
        message = await self.websocket.recv()
        return json.loads(message)
    
    async def ping(self):
        """Send ping"""
        if not self.websocket:
            raise RuntimeError("Not connected")
        
        await self.websocket.send(json.dumps({"action": "ping"}))
        return await self.receive()
    
    async def close(self):
        """Close connection"""
        if self.websocket:
            await self.websocket.close()
            self.websocket = None


class TradingClient:
    """
    Main AI Trading Evolution client
    
    Example:
        async with TradingClient(api_key="your-key") as client:
            # Get market data
            data = await client.market.get_data("BTC/USDT")
            
            # Execute strategy
            signal = await client.strategy.execute("grid", "BTC/USDT")
            
            # Run analytics
            sentiment = await client.analytics.sentiment("BTC/USDT")
    """
    
    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_key: Optional[str] = None,
        timeout: int = 30
    ):
        self._base_client = BaseClient(base_url, api_key, timeout)
        
        # API endpoints
        self.market = MarketAPI(self._base_client)
        self.strategy = StrategyAPI(self._base_client)
        self.analytics = AnalyticsAPI(self._base_client)
    
    async def __aenter__(self):
        await self._base_client.connect()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self._base_client.close()
    
    async def connect(self):
        """Connect to API"""
        await self._base_client.connect()
    
    async def close(self):
        """Close connection"""
        await self._base_client.close()
    
    async def health(self) -> Dict[str, Any]:
        """Get API health status"""
        return await self._base_client.get("/health")
    
    async def metrics(self) -> Dict[str, Any]:
        """Get performance metrics"""
        return await self._base_client.get("/metrics")
    
    def websocket(self) -> WebSocketClient:
        """Get WebSocket client"""
        ws_url = self._base_client.base_url.replace("http://", "ws://").replace("https://", "wss://")
        return WebSocketClient(ws_url)


# =============================================================================
# Synchronous wrapper (optional)
# =============================================================================

class SyncTradingClient:
    """
    Synchronous wrapper for TradingClient
    
    Example:
        client = SyncTradingClient(api_key="your-key")
        data = client.market.get_data("BTC/USDT")
        client.close()
    """
    
    def __init__(self, *args, **kwargs):
        self._async_client = TradingClient(*args, **kwargs)
        self._loop = asyncio.get_event_loop()
    
    def _run(self, coro):
        """Run async coroutine"""
        return self._loop.run_until_complete(coro)
    
    def __enter__(self):
        self._run(self._async_client.connect())
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self._run(self._async_client.close())
    
    @property
    def market(self):
        """Market API"""
        class SyncMarketAPI:
            def __init__(self, async_api, loop):
                self._async_api = async_api
                self._loop = loop
            
            def get_data(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.get_data(*args, **kwargs)
                )
            
            def list_symbols(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.list_symbols(*args, **kwargs)
                )
        
        return SyncMarketAPI(self._async_client.market, self._loop)
    
    @property
    def strategy(self):
        """Strategy API"""
        class SyncStrategyAPI:
            def __init__(self, async_api, loop):
                self._async_api = async_api
                self._loop = loop
            
            def execute(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.execute(*args, **kwargs)
                )
            
            def list(self):
                return self._loop.run_until_complete(
                    self._async_api.list()
                )
        
        return SyncStrategyAPI(self._async_client.strategy, self._loop)
    
    @property
    def analytics(self):
        """Analytics API"""
        class SyncAnalyticsAPI:
            def __init__(self, async_api, loop):
                self._async_api = async_api
                self._loop = loop
            
            def analyze(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.analyze(*args, **kwargs)
                )
            
            def sentiment(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.sentiment(*args, **kwargs)
                )
            
            def risk_scoring(self, *args, **kwargs):
                return self._loop.run_until_complete(
                    self._async_api.risk_scoring(*args, **kwargs)
                )
            
            def list_types(self):
                return self._loop.run_until_complete(
                    self._async_api.list_types()
                )
        
        return SyncAnalyticsAPI(self._async_client.analytics, self._loop)
    
    def health(self):
        """Get health status"""
        return self._run(self._async_client.health())
    
    def metrics(self):
        """Get metrics"""
        return self._run(self._async_client.metrics())
    
    def close(self):
        """Close connection"""
        self._run(self._async_client.close())
